test/enginetest.c

```diff
@@ -175,6 +176,177 @@ static int test_engines(void)
         ENGINE_free(block[loop]);
     return to_return;
 }
+
+/* Test EVP_PKEY method */
+static EVP_PKEY_METHOD *test_rsa = NULL;
+
+static int called_encrypt = 0;
+
+/* Test function to check operation has been redirected */
+static int test_encrypt(EVP_PKEY_CTX *ctx, unsigned char *sig,
+                        size_t *siglen, const unsigned char *tbs, size_t tbslen)
+{
+    called_encrypt = 1;
+    return 1;
+}
+
+static int test_pkey_meths(ENGINE *e, EVP_PKEY_METHOD **pmeth,
+                           const int **pnids, int nid)
+{
+    static const int rnid = EVP_PKEY_RSA;
+    if (pmeth == NULL) {
+        *pnids = &rnid;
+        return 1;
+    }
+
+    if (nid == EVP_PKEY_RSA) {
+        *pmeth = test_rsa;
+        return 1;
+    }
+
+    *pmeth = NULL;
+    return 0;
+}
+
+/* Return a test EVP_PKEY value */
+
+static EVP_PKEY *get_test_pkey(void)
+{
+    static unsigned char n[] =
+        "\x00\xAA\x36\xAB\xCE\x88\xAC\xFD\xFF\x55\x52\x3C\x7F\xC4\x52\x3F"
+        "\x90\xEF\xA0\x0D\xF3\x77\x4A\x25\x9F\x2E\x62\xB4\xC5\xD9\x9C\xB5"
+        "\xAD\xB3\x00\xA0\x28\x5E\x53\x01\x93\x0E\x0C\x70\xFB\x68\x76\x93"
+        "\x9C\xE6\x16\xCE\x62\x4A\x11\xE0\x08\x6D\x34\x1E\xBC\xAC\xA0\xA1"
+        "\xF5";
+    static unsigned char e[] = "\x11";
+
+    RSA *rsa;
+    EVP_PKEY *pk;
+
+    rsa = RSA_new();
+
+    pk = EVP_PKEY_new();
+
+    if (rsa == NULL || pk == NULL)
+        return NULL;
+
+    EVP_PKEY_assign_RSA(pk, rsa);
+
+    if (!RSA_set0_key(rsa, BN_bin2bn(n, sizeof(n)-1, NULL),
+                      BN_bin2bn(e, sizeof(e)-1, NULL), NULL)) {
+        EVP_PKEY_free(pk);
+        return NULL;
+    }
+
+    return pk;
+}
+
+static int test_redirect(void)
+{
+    const unsigned char pt[] = "Hello World\n";
+    unsigned char *tmp = NULL;
+    size_t len;
+    EVP_PKEY_CTX *ctx = NULL;
+    ENGINE *e = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    int to_return = 0;
+
+    if (!TEST_ptr(pkey = get_test_pkey()))
+        goto err;
+
+    len = EVP_PKEY_size(pkey);
+    if (!TEST_ptr(tmp = OPENSSL_malloc(len)))
+        goto err;
+
+    if (!TEST_ptr(ctx = EVP_PKEY_CTX_new(pkey, NULL)))
+        goto err;
+    TEST_info("EVP_PKEY_encrypt test: no redirection");
+    /* Encrypt some data: should succeed but not be redirected */
+    if (!TEST_true(EVP_PKEY_encrypt_init(ctx) > 0)
+            || !TEST_true(EVP_PKEY_encrypt(ctx, tmp, &len, pt, sizeof(pt)) > 0)
```

Comment:
TEST_int_gt() instead of TEST_true() in these two lines and other places in this function?

Reply:
Addressed this and all the other cases (unless I missed one) in latest commit.